### PR TITLE
feat: obtain and keep chainLock in sync with core

### PR DESCRIPTION
**Pull request by Alex-Werner**
## Issue being fixed or feature implemented

In order to provide the latest chainlock to platform, we need to get best chainlock from RPC and maintain ZMQ connectivity to keep it in sync with additional chainlocks.
Since ChainLock could come before block, we need to make sure that block is present as well.

## What was done?
- feat: LatestCoreChainLock maintaining chainlock value
- feat: on initChain request chainlock via RPC
- feat: on initChain keep chainlock in sync via ZMQ

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by Alex-Werner** [comment]
May be we could handle our ZMQClient in a new repository, the same way we manage RPCClient in dashd-rpc (we would then have a dashd-zmq) ?

File: lib/abci/handlers/endBlockHandlerFactory.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * End block ABCI handler
+ *
+ * @param {BlockchainState} blockchainState
+ * @param {BaseLogger} logger
+ *
+ * @return {endBlockHandler}
+ */
+function endBlockHandlerFactory(
```
Comment (shumkov) [inline lib/abci/handlers/endBlockHandlerFactory.js:15]: This is out of scope of this task. Here we just need to obtain ChainLock and store it in LatestCoreChainLock class

File: lib/abci/handlers/initChainHandlerFactory.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+   * @typedef endBlockHandler
+   *
+   * @param {abci.RequestInitChain} request
+   * @return {Promise<abci.ResponseInitChain>}
+   */
+  async function initChainHandler() {
+    logger.info('Init chain');
+
+    // We need to retrieve latest chainlock from our fully synced DashCore instance
+    const rpcBestChainLockResponse = await coreRpcClient.getBestChainLock();
```
Comment (shumkov) [inline lib/abci/handlers/initChainHandlerFactory.js:36]: As descrived in taks we need to obtain initial chainlock before we start ABCI server so it should be implemented in scripts/abci

File: lib/abci/handlers/endBlockHandlerFactory.js (hunk 1 above)
Comment (Alex-Werner) [reply lib/abci/handlers/endBlockHandlerFactory.js:15]: Another task required endBlock and this PR is built on top of that.

File: lib/core/LatestCoreChainLock.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+class LatestCoreChainLock {
+  /**
+   *
+   * @param {string} [chainLock]
+   * @param {LatestCoreChainLock}
+   */
+  constructor(chainLock) {
+    this.chainLock = chainLock || null;
```
Comment (shumkov) [inline lib/core/LatestCoreChainLock.js:8]: We use undefined for now everywhere.

File: lib/core/LatestCoreChainLock.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+class LatestCoreChainLock {
+  /**
+   *
+   * @param {string} [chainLock]
```
Comment (shumkov) [inline lib/core/LatestCoreChainLock.js:4]: Let's use ChainLock primitive?

File: lib/core/LatestCoreChainLock.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+   * @param {LatestCoreChainLock}
+   */
+  constructor(chainLock) {
+    this.chainLock = chainLock || null;
+  }
+
+  /**
+   * Update latest chainlock
+   *
+   * @param {string} chainLock
```
Comment (shumkov) [inline lib/core/LatestCoreChainLock.js:14]: ChainLock

File: lib/abci/handlers/endBlockHandlerFactory.js (hunk 1 above)
Comment (shumkov) [reply lib/abci/handlers/endBlockHandlerFactory.js:15]: Yeah. Probably we could just remove handlers from this branch? :))

File: lib/core/ZMQClient.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+  subscribe(roomName) {
+    if (!Object.keys(ZMQClient.ROOMS).includes(roomName)) {
+      throw new Error('Invalid roomName');
+    }
+    this.socket.subscribe(roomName);
+  }
+}
+
+ZMQClient.ROOMS = {
```
Comment (shumkov) [inline lib/core/ZMQClient.js:35]: TOPICS

File: test/unit/core/LatestCoreChainLock.spec.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
+const LatestCoreChainLock = require('../../../lib/core/LatestCoreChainLock');
+
+describe('LatestCoreChainLock', () => {
+  beforeEach(() => {
+  });
+
+  describe('instantiation', () => {
```
Comment (shumkov) [inline test/unit/core/LatestCoreChainLock.spec.js:7]: #constructor

File: test/unit/core/LatestCoreChainLock.spec.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    it('should instantiates', () => {
+      const latestCoreChainLock = new LatestCoreChainLock();
+      expect(latestCoreChainLock.chainLock).to.equal(null);
+      const latestCoreChainLockWithValue = new LatestCoreChainLock('someValue');
+      expect(latestCoreChainLockWithValue.chainLock).to.equal('someValue');
+    });
+  });
+  describe('usage', () => {
```
Comment (shumkov) [inline test/unit/core/LatestCoreChainLock.spec.js:15]: #update

File: lib/abci/handlers/endBlockHandlerFactory.js (hunk 1 above)
Comment (Alex-Werner) [reply lib/abci/handlers/endBlockHandlerFactory.js]: Removed !

File: lib/core/waitForCoreChainLockSyncFactory.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    });
+
+    // Because a Chainlock may happen before its block, we also subscribe to rawblock
+    coreZMQClient.subscribe(ZMQClient.TOPICS.hashblock);
+    logger.info('Subscribe to hashblock ZMQ room');
+
+    // We need to retrieve latest chainlock from our fully synced DashCore instance
+    try {
+      const rpcBestChainLockResponse = await coreRpcClient.getBestChainLock();
+      const chainlock = new ChainLock(rpcBestChainLockResponse.result);
```
Comment (shumkov) [inline lib/core/waitForCoreChainLockSyncFactory.js:51]: What if Chain Lock is not present?

File: lib/util/ensureBlock.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+const ZMQClient = require('@dashevo/dashd-zmq');
+const waitForEvent = require('./waitForEvent');
+
+async function ensureBlock(socketClient, rpcClient, hash) {
```
Comment (shumkov) [inline lib/util/ensureBlock.js:4]: It looks more like business logic than infrastructure. I would move it to `lib/core` too

File: scripts/abci.js [hunk 11]
```diff
@@ -27,6 +27,10 @@ const errorHandler = require('../lib/errorHandler');
     );
   });
 
+  logger.info('Synchronizing latestChainlock');
```
Comment (shumkov) [inline scripts/abci.js:30]: "Obtaining the latest Core ChainLock"?

File: lib/util/waitForEvent.js [hunk 12]
```diff
@@ -0,0 +1,18 @@
+/**
+ * Asynchronously wait for an event to be fired.
+ * @param {EventEmitter} emitter - emitter of the event
+ * @param {String} eventName - eventName to wait for
+ * @param {String} [value] - Optionally will verify value to equal response
+ * @return {Promise<any>} The promise to await on.
+ */
+async function waitForEvent(emitter, eventName, value) {
```
Comment (shumkov) [inline lib/util/waitForEvent.js:8]: `value = undefined`

File: lib/core/waitForCoreChainLockSyncFactory.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+    await coreZMQClient.connect();
+    // listens to ChainLock zmq zmqpubrawchainlocksig event and updates the latestCoreChainLock.
+    // If we lost connection with Core (heartbeats with retry policy) we need to exit with fatal
+    // In case of regtest fallback we need to listen for new blocks using hashblock zmq event
+    coreZMQClient.subscribe(ZMQClient.TOPICS.rawchainlock);
+    logger.info('Subscribe to rawchainlock ZMQ room');
+
+    coreZMQClient.on(ZMQClient.TOPICS.rawchainlock, async (rawChainLockMessage) => {
+      const socketChainLock = new ChainLock(rawChainLockMessage);
+      logger.info(`Received new chainlock ${socketChainLock.toString()}`);
```
Comment (shumkov) [inline lib/core/waitForCoreChainLockSyncFactory.js:39]: I would make it `debug`.

File: lib/core/waitForCoreChainLockSyncFactory.js [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+   *
+   * @returns {Promise<LatestCoreChainLock>}
+   */
+  async function waitForCoreChainLockSync() {
+    await coreZMQClient.connect();
+    // listens to ChainLock zmq zmqpubrawchainlocksig event and updates the latestCoreChainLock.
+    // If we lost connection with Core (heartbeats with retry policy) we need to exit with fatal
+    // In case of regtest fallback we need to listen for new blocks using hashblock zmq event
+    coreZMQClient.subscribe(ZMQClient.TOPICS.rawchainlock);
+    logger.info('Subscribe to rawchainlock ZMQ room');
```
Comment (shumkov) [inline lib/core/waitForCoreChainLockSyncFactory.js:35]: Do you think it could help?

File: lib/core/waitForCoreChainLockSyncFactory.js [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    coreZMQClient.on(ZMQClient.TOPICS.rawchainlock, async (rawChainLockMessage) => {
+      const socketChainLock = new ChainLock(rawChainLockMessage);
+      logger.info(`Received new chainlock ${socketChainLock.toString()}`);
+      await ensureBlock(coreZMQClient, coreRpcClient, socketChainLock.blockHash);
+      latestCoreChainLock.update(socketChainLock);
+    });
+
+    // Because a Chainlock may happen before its block, we also subscribe to rawblock
+    coreZMQClient.subscribe(ZMQClient.TOPICS.hashblock);
+    logger.info('Subscribe to hashblock ZMQ room');
```
Comment (shumkov) [inline lib/core/waitForCoreChainLockSyncFactory.js:46]: Do you think it could help?

File: lib/core/waitForCoreChainLockSyncFactory.js [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+      const chainlock = new ChainLock(rpcBestChainLockResponse.result);
+
+      await ensureBlock(coreZMQClient, coreRpcClient, chainlock.blockHash);
+      latestCoreChainLock.update(chainlock);
+    } catch (e) {
+      // Unable to find any chainlock
+      if (e.code === -32603) {
+        logger.info('Waiting for a new chainlock');
+        // We need to wait for new chainlock from ZMQ socket
+        await waitForEvent(coreZMQClient, ZMQClient.TOPICS.rawchainlock);
```
Comment (shumkov) [inline lib/core/waitForCoreChainLockSyncFactory.js:60]: It waits when an event happens but it doesn't wait until the listener above will finish. I feel we need a slightly different logic here.

File: lib/core/waitForCoreChainLockSyncFactory.js (hunk 15 above)
Comment (Alex-Werner) [reply lib/core/waitForCoreChainLockSyncFactory.js:46]: As `info` may be not so much, as a `trace` yeah, it's a dependency, if you have an issue at this point, trace might highlight the where (i'm pretty pro-verbose as trace tbh).

File: lib/core/waitForCoreChainLockSyncFactory.js (hunk 9 above)
Comment (Alex-Werner) [reply lib/core/waitForCoreChainLockSyncFactory.js:51]: We actually need to handle the error case in order to provide a fallback way for syncing chainLock, else what we just got thrown out of there by the error (which is not what we want as the error is legit).

File: lib/core/waitForCoreChainLockSyncFactory.js (hunk 16 above)
Comment (Alex-Werner) [reply lib/core/waitForCoreChainLockSyncFactory.js:60]: When you end up here, it means that rpcBestChainLockResponse has errored into a -32603 error code, so you never `ensureBlock`, when you got to `ensureBlock`, you just throw error (because error won't be -32603), so both condition are not met in the same flow.

File: scripts/abci.js (hunk 11 above)
Comment (Alex-Werner) [reply scripts/abci.js:30]: Fixed

File: lib/core/waitForCoreChainLockSyncFactory.js (hunk 15 above)
Comment (shumkov) [reply lib/core/waitForCoreChainLockSyncFactory.js:46]: Sure! let's change the logging level for this output!

File: lib/core/waitForCoreChainLockSyncFactory.js (hunk 14 above)
Comment (shumkov) [reply lib/core/waitForCoreChainLockSyncFactory.js:35]: Agreed to use `trace`

File: lib/core/waitForCoreChainLockSyncFactory.js (hunk 15 above)
Comment (shumkov) [reply lib/core/waitForCoreChainLockSyncFactory.js:46]: Agreed to use `trace`

File: lib/core/waitForCoreChainLockSyncFactory.js [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+   * Check that chainLock is synced
+   *
+   * @typedef waitForCoreChainLockSync
+   *
+   * @returns {Promise<LatestCoreChainLock>}
+   */
+  async function waitForCoreChainLockSync() {
+    await coreZMQClient.connect();
+
+    // By default will try to reconnect so we just log when this happen
```
Comment (antouhou) [inline lib/core/waitForCoreChainLockSyncFactory.js:33]: Can you send me a link to the code that reconnects on the `disconnect` event?

File: test/unit/core/LatestCoreChainLock.spec.js [hunk 18]
```diff
@@ -0,0 +1,22 @@
+const LatestCoreChainLock = require('../../../lib/core/LatestCoreChainLock');
+
+describe('LatestCoreChainLock', () => {
+  beforeEach(() => {
+  });
+
+  describe('#constructor', () => {
+    it('should instantiates', () => {
```
Comment (antouhou) [inline test/unit/core/LatestCoreChainLock.spec.js:8]: instantiate

File: lib/core/waitForCoreChainLockSyncFactory.js (hunk 17 above)
Comment (Alex-Werner) [reply lib/core/waitForCoreChainLockSyncFactory.js:33]: The best way is to test it. Subscribe to connect:delay, connect, and reconnect. 
Else what look up zmq.Observer.Events for the events or zmq.Subscriber for the reconnect. 
The JS doc in ZMQ-dashd (or it's docsify) also documents it.

File: test/unit/core/LatestCoreChainLock.spec.js (hunk 18 above)
Comment (Alex-Werner) [reply test/unit/core/LatestCoreChainLock.spec.js:8]: Fix ! 👍

File: lib/core/waitForCoreChainLockSyncFactory.js (hunk 17 above)
Comment (Alex-Werner) [reply lib/core/waitForCoreChainLockSyncFactory.js:33]: - [SocketMonitor that allow us to get .on() on those events](https://github.com/dashevo/js-dashd-zmq/blob/master/lib/ZMQClient/methods/connect.js#L26)
- [ReconnectInterval as part of passed options from connect() method](https://github.com/zeromq/zeromq.js/blob/master/src/index.ts#L524)
- [As its binding on libzmq, the socks_connecter file display the reconnect implementation on connect:error](https://github.com/zeromq/libzmq/blob/master/src/socks_connecter.cpp#L273)

Comment (Alex-Werner) [reply lib/core/waitForCoreChainLockSyncFactory.js:33]: So disconnection is default, but settings interval to -1 will overwrite default reconnection attempts.

File: lib/core/waitForCoreChainLockSyncFactory.js [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+    coreZMQClient.on('disconnect', logger.error);
```
Comment (shumkov) [inline lib/core/waitForCoreChainLockSyncFactory.js:34]: Let’s log it with `trace`?

File: lib/core/waitForCoreChainLockSyncFactory.js [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+
+    // When socket monitoring ends
+    coreZMQClient.on('end', logger.error);
```
Comment (shumkov) [inline lib/core/waitForCoreChainLockSyncFactory.js:37]: Don’t we need to call errorHandler here to shutdown Drive?

File: lib/core/waitForCoreChainLockSyncFactory.js (hunk 17 above)
Comment (shumkov) [reply lib/core/waitForCoreChainLockSyncFactory.js:33]: how many times we try to reconnect?

File: lib/core/waitForCoreChainLockSyncFactory.js (hunk 14 above)
Comment (shumkov) [reply lib/core/waitForCoreChainLockSyncFactory.js:45]: Should we change to `trace`?

File: test/unit/core/ensureBlock.spec.js [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+    rpcClient = {
+      getBlock: async () => {
+        const err = new Error();
+        err.code = -6;
+        err.message = 'Another error';
+        throw err;
+      },
+    };
+    const p = ensureBlock(socketClient, rpcClient, hash);
+    p.should.be.rejectedWith(Error, 'Another error');
```
Comment (shumkov) [inline test/unit/core/ensureBlock.spec.js:51]: We use expect usually

File: test/unit/core/ensureBlock.spec.js [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+  const socketClient = new EventEmitter();
+  let rpcClient;
+  it('should ensure a block exist before returning promise', (done) => {
+    rpcClient = {
+      getBlock: async () => true,
+    };
+    ensureBlock(socketClient, rpcClient, hash)
+      .then(() => done());
+  });
+  it('should wait for block if not found before returning promise', (done) => {
```
Comment (shumkov) [inline test/unit/core/ensureBlock.spec.js:23]: We need to mock `waitForEvent` I guess. We could move it to dependencies then.

File: lib/createDIContainer.js [hunk 23]
```diff
@@ -132,10 +137,13 @@ async function createDIContainer(options) {
     dataContractLevelDBFile: asValue(options.DATA_CONTRACT_LEVEL_DB_FILE),
     documentMongoDBPrefix: asValue(options.DOCUMENT_MONGODB_DB_PREFIX),
     documentMongoDBUrl: asValue(options.DOCUMENT_MONGODB_URL),
+    chainLock: asValue(undefined),
```
Comment (shumkov) [inline lib/createDIContainer.js:140]: What for?

File: lib/core/waitForCoreChainLockSyncFactory.js (hunk 17 above)
Comment (Alex-Werner) [reply lib/core/waitForCoreChainLockSyncFactory.js:33]: We can have a backoff mechanism with an intervalMax : (see : [ZMQ_RECONNECT_IVL_MAX](http://api.zeromq.org/4-2:zmq-setsockopt#toc38) and [reconnectMaxInterval](https://zeromq.github.io/zeromq.js/classes/subscriber.html#reconnectmaxinterval))
But no limit by default, but implementation of a reconnect stop is actually pretty trivial as each connection tries is emited and listenable, a simple counter that would call .close() after X retries is something I can implement. 
I just need to know which rules we go for, what is our max, etc...

File: lib/createDIContainer.js (hunk 23 above)
Comment (Alex-Werner) [reply lib/createDIContainer.js:140]: I got an error with createDIContainer when this is not existant where it fails to build dep tree.

Comment (shumkov) [reply lib/createDIContainer.js:140]: I see. You can just use `asFunction`  for `latestCoreChainLock` and instantiate the class by yourself or remove the constructor at all.

File: test/unit/core/ensureBlock.spec.js (hunk 21 above)
Comment (Alex-Werner) [reply test/unit/core/ensureBlock.spec.js]: Fixed

File: lib/core/waitForCoreChainLockSyncFactory.js (hunk 14 above)
Comment (shumkov) [reply lib/core/waitForCoreChainLockSyncFactory.js:48]: trace, like in 73rd line?

File: lib/core/ensureBlock.js [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * @param {EventEmitter} socketClient
+ * @param {RpcClient} rpcClient
+ * @param {string} hash
+ * @return {Promise<void>}
+ */
+async function ensureBlock(socketClient, rpcClient, hash) {
+  const eventPromise = new Promise((resolve) => {
+    socketClient.on(ZMQClient.TOPICS.hashblock, (response) => {
+      if (hash === undefined || hash === response) {
```
Comment (shumkov) [inline lib/core/ensureBlock.js:13]: Hash can be undefined?

File: lib/core/ensureBlock.js [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ *
+ * @param {EventEmitter} socketClient
+ * @param {RpcClient} rpcClient
+ * @param {string} hash
+ * @return {Promise<void>}
+ */
+async function ensureBlock(socketClient, rpcClient, hash) {
+  const eventPromise = new Promise((resolve) => {
+    socketClient.on(ZMQClient.TOPICS.hashblock, (response) => {
```
Comment (shumkov) [inline lib/core/ensureBlock.js:12]: I guess you need to remove the listener when you get the block.

File: lib/core/fallbackChainLockSyncFactory.js [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+    const signature = Buffer.alloc(32);
+
+    await coreZMQClient.connect();
+
+    // By default will try to reconnect so we just log when this happen
+    coreZMQClient.on('disconnect', logger.trace);
+
+    // When socket monitoring ends
+    coreZMQClient.on('end', (caughtError) => {
+      const error = new Error(`Can't continue without chainLock: ${caughtError.message}`);
```
Comment (shumkov) [inline lib/core/fallbackChainLockSyncFactory.js:37]: `Lost connection with Core: ${caughtError.message}`

File: lib/core/fallbackChainLockSyncFactory.js [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+
+      errorHandler(error);
+    });
+
+    coreZMQClient.subscribe(ZMQClient.TOPICS.hashblock);
+
+    logger.info('Subscribe to hashblock ZMQ room');
```
Comment (shumkov) [inline lib/core/fallbackChainLockSyncFactory.js:44]: trace

File: lib/core/fallbackChainLockSyncFactory.js [hunk 28]
```diff
@@ -0,0 +1,89 @@
...
+      latestCoreChainLock.update(socketChainLock);
+
+      if (resolveFirstBlockFromZMQPromise) {
+        resolveFirstBlockFromZMQPromise();
+        resolveFirstBlockFromZMQPromise = null;
+      }
+    });
+
+    const { result: rpcBestBlockHash } = await coreRpcClient.getBestBlockHash();
+    const { result: rpcBestBlock } = await coreRpcClient.getBlock(rpcBestBlockHash);
```
Comment (shumkov) [inline lib/core/fallbackChainLockSyncFactory.js:68]: What does it respond in case if there are no blocks? I mean how hash does look like?

File: lib/core/waitForCoreChainLockSyncFactory.js (hunk 14 above)
Comment (shumkov) [reply lib/core/waitForCoreChainLockSyncFactory.js:52]: Must be trace, not `info`

File: lib/core/waitForCoreChainLockSyncFactory.js [hunk 29]
```diff
@@ -0,0 +1,107 @@
...
+    });
+
+    coreZMQClient.on(ZMQClient.TOPICS.rawchainlock, async (rawChainLockMessage) => {
+      const socketChainLock = new ChainLock(rawChainLockMessage);
+
+      await ensureBlock(coreZMQClient, coreRpcClient, socketChainLock.blockHash);
+
+      latestCoreChainLock.update(socketChainLock);
+
+      logger.debug(`Received new chainlock ${socketChainLock.toString()}`);
```
Comment (shumkov) [inline lib/core/waitForCoreChainLockSyncFactory.js:66]: logger.debug(socketChainLock.toJSON(), 'Received new chainlock')

File: lib/core/fallbackChainLockSyncFactory.js (hunk 28 above)
Comment (shuplenkov) [reply lib/core/fallbackChainLockSyncFactory.js:68]: genesis block with height = 0

File: lib/core/fallbackChainLockSyncFactory.js [hunk 30]
```diff
@@ -0,0 +1,89 @@
...
+    coreZMQClient.on(ZMQClient.TOPICS.hashblock, async (blockHash) => {
+      const { result: block } = await coreRpcClient.getBlock(blockHash);
+      const socketChainLock = new ChainLock({
+        height: block.height,
+        blockHash,
+        signature,
+      });
+
+      latestCoreChainLock.update(socketChainLock);
+
```
Comment (shumkov) [inline lib/core/fallbackChainLockSyncFactory.js:60]: logger.debug(socketChainLock.toJSON(), 'Updated latestCoreChanLock');

File: scripts/abci.js [hunk 31]
```diff
@@ -20,12 +20,25 @@ const errorHandler = require('../lib/errorHandler');
...
+      logger.info(
+        `waiting for Core to finish sync ${currentBlockHeight}/${currentHeaderNumber}...`,
+      );
+    });
+
+    logger.info('Obtaining the latest Core ChainLock...');
+    const waitForCoreChainLockSync = container.resolve('waitForCoreChainLockSync');
+    await waitForCoreChainLockSync();
+  } else {
+    const fallbackChainLockSync = container.resolve('fallbackChainLockSync');
```
Comment (shumkov) [inline scripts/abci.js:39]: waitForCoreChainLockSyncFallback

Comment (shumkov) [reply scripts/abci.js:39]: logger.info('Obtaining the latest core block for chain lock sync fallback...');

File: lib/core/fallbackChainLockSyncFactory.js (hunk 28 above)
Comment (shumkov) [reply lib/core/fallbackChainLockSyncFactory.js:68]: Got it.

Comment (shumkov) [reply lib/core/fallbackChainLockSyncFactory.js:68]: Yeah, it's better to use the next block.

File: lib/core/ensureBlock.js [hunk 32]
```diff
@@ -2,15 +2,20 @@ const ZMQClient = require('@dashevo/dashd-zmq');
...
- * @param {EventEmitter} socketClient
+ * @param {ZMQClient} zmqClient
  * @param {RpcClient} rpcClient
  * @param {string} hash
  * @return {Promise<void>}
  */
-async function ensureBlock(socketClient, rpcClient, hash) {
+async function ensureBlock(zmqClient, rpcClient, hash) {
+  // Because a ChainLock may happen before its block, we also subscribe to rawblock
+  zmqClient.subscribe(ZMQClient.TOPICS.hashblock);
```
Comment (shumkov) [inline lib/core/ensureBlock.js:12]: Don't we need to subscribe in `waitFor...`?

File: lib/core/ensureBlock.js [hunk 33]
```diff
@@ -2,15 +2,20 @@ const ZMQClient = require('@dashevo/dashd-zmq');
...
+
   const eventPromise = new Promise((resolve) => {
-    socketClient.on(ZMQClient.TOPICS.hashblock, (response) => {
-      if (hash === undefined || hash === response) {
+    zmqClient.on(ZMQClient.TOPICS.hashblock, (response) => {
+      if (hash === response) {
+        zmqClient.removeAllListeners(ZMQClient.TOPICS.hashblock);
```
Comment (shumkov) [inline lib/core/ensureBlock.js:17]: It could lead to problems when other part of the code that will use zmqClient. Let's just remove this exact listener?

File: package.json [hunk 34]
```diff
@@ -89,6 +90,7 @@
     "pino": "^6.4.0",
     "pino-pretty": "^4.0.3",
     "setimmediate": "^1.0.5",
-    "through2": "^3.0.1"
+    "through2": "^3.0.1",
+    "zeromq": "^6.0.0-beta.6"
```
Comment (shumkov) [inline package.json:94]: I guess we don't need it since we are using `@dashevo/dashd-zmq`

File: test/integration/core/waitForCoreSyncFactory.spec.js [hunk 35]
```diff
@@ -62,26 +62,4 @@ describe('waitForCoreSyncFactory', function main() {
     expect(currentBlockHeight).to.equal(1000);
     expect(currentHeadersNumber).to.equal(1000);
   });
-
-  it('shouldn\'t wait if Dash Core in regtest mode without peers', async () => {
```
Comment (shumkov) [inline test/integration/core/waitForCoreSyncFactory.spec.js:66]: We need to move it to `detectStandaloneRegtestModeFactory.spec.js`

**Review by shumkov — APPROVED** [review]
Good job, guys! 👍
